### PR TITLE
[FEATURE] Add `PluginConfig`

### PR DESCRIPTION
**Pull request by mariahjrogers**
## Pull request checklist

Before submitting your PR, please review the following checklist:

- [ ] Consider adding a unit test if your PR resolves an issue.

- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

### Breaking changes

- [ ] Check if this pull request introduces a breaking change

## What does this implement/fix? Explain your changes.

A description of the problem you're trying to solve.
Any relevant logs, error output, etc.
An overview of the suggested solution
If the feature changes current behavior, reasons why your solution is better.

## Does this close any currently open issues?
…

## Any other comments?
…

## Where has this been tested?

**Operating System:** …

**Platform:** …

**Target Platform:** …

### CODE REVIEW

File: palm/plugins/dbt/commands/cmd_dbt-config.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+@click.command("dbt-config")
+@click.pass_obj
+def cli(environment):
+    """Set up your palm dbt configuration"""
+    click.echo("Setting up your palm dbt configuration")
+    config = {}
+
+    has_prod_artifacts = click.confirm("Do you have production dbt artifacts saved locally?")
+    if has_prod_artifacts:
+        config['dbt_artifacts_prod'] = click.prompt("Prod artifacts location:")
```
Comment (jakeberesford-palmetto) [inline palm/plugins/dbt/commands/cmd_dbt-config.py:14]: `type=Path()` will improve both of these prompts (`from pathlib import Path`)

File: palm/plugins/dbt/plugin_config.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+from pathlib import Path
+from click import secho
+from typing import Optional
+import yaml
+
+
+class PluginConfig:
+
+    def __init__(self, project_path: Optional["Path"] = Path.cwd()):
+        self.project_root = project_path
```
Comment (jakeberesford-palmetto) [inline palm/plugins/dbt/plugin_config.py:10]: Unnecessary variable assignment, this line can be removed

File: palm/plugins/dbt/plugin_config.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        if not key:
+            raise ValueError('Local artifacts path not found, run `palm dbt-config`')
+        return key
+
+    @property
+    def dbt_artifacts_prod(self) -> str:
+        secret = self.config.get('dbt_artifacts_prod')
+        if not secret:
+            raise ValueError(
+                'Prod artifacts path not found, run `palm dbt-config`')
```
Comment (jakeberesford-palmetto) [inline palm/plugins/dbt/plugin_config.py:38]: nit, closing brace should go to the next line
```suggestion
                'Prod artifacts path not found, run `palm dbt-config`'
            )
```

File: palm/plugins/dbt/dbt_palm_utils.py [hunk 4]
```diff
@@ -51,6 +52,7 @@ def dbt_env_vars(branch: str) -> Dict:
         'PDP_DEV_SCHEMA': _generate_schema_from_branch(branch),
         'PDP_ENV': 'DEVELOPMENT',  # Deprecated - this will be removed!
         'PALM_DBT_ENV': 'DEVELOPMENT',
+        'DBT_'
```
Comment (jakeberesford-palmetto) [inline palm/plugins/dbt/dbt_palm_utils.py:55]: remove, add the env vars only in the commands you want, and only when you need them!

**Review by jakeberesford-palmetto — COMMENTED** [review]
Good start, this will be pretty cool!

File: palm/plugins/dbt/commands/cmd_run.py [hunk 5]
```diff
@@ -1,6 +1,7 @@
 import click
 from typing import Optional, Tuple
 from palm.plugins.dbt.dbt_palm_utils import shell_options, dbt_env_vars
+from palm.plugins.dbt.plugin_config import PluginConfig
```
Comment (jakeberesford-palmetto) [inline palm/plugins/dbt/commands/cmd_run.py:4]: Unused import, this should be removed for now. (I know we have plans to use it in a follow-up PR)

**Review by jakeberesford-palmetto — COMMENTED** [review]
A couple of small changes are needed, but this is very close! Excited about the capabilities this will unlock!

File: palm/plugins/dbt/plugin_config.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    @property
+    def dbt_artifacts_local(self) -> str:
+        key = self.config.get('dbt_artifacts_local')
+        if not key:
+            raise ValueError('Local artifacts path not found, run `palm dbt-config`')
+        return key
+
+    @property
+    def dbt_artifacts_prod(self) -> str:
+        secret = self.config.get('dbt_artifacts_prod')
```
[the remark this answers is not in the document]
Comment (mariahjrogers) [reply palm/plugins/dbt/plugin_config.py]: Very odd, I don't know why this is named secret. Changed it to key!

File: palm/plugins/dbt/commands/cmd_dbt-config.py [hunk 7]
```diff
@@ -0,0 +1,29 @@
+import click
+from pathlib import Path
+from palm.plugins.dbt.plugin_config import PluginConfig
+
+
+@click.command("dbt-config")
```
Comment (mariahjrogers) [inline palm/plugins/dbt/commands/cmd_dbt-config.py:6]: @jakeberesford-palmetto question... what if it were just called *config* 🤔

I would love for it to be run as `palm dbt config` which feels so much cleaner, the issue here is that `palm dbt` is going to become the passthrough command, and not a reference to the dbt plugin itself. Also, `config` alone would be overwritten/conflicted with other plugins' `config` commands or local config commands.

Comment (jakeberesford-palmetto) [reply palm/plugins/dbt/commands/cmd_dbt-config.py:6]: I like `config` :)

Comment (mariahjrogers) [reply palm/plugins/dbt/commands/cmd_dbt-config.py:6]: But what about multiple plugins with `config` commands?

Comment (jakeberesford-palmetto) [reply palm/plugins/dbt/commands/cmd_dbt-config.py:6]: That gets gnarly... depends on the plugin load order, the last loaded would be the one to run... not great.

We probably want to think about adding `PluginConfig` as a feature of `BasePlugin` and designing a solution that will work well for generating the config when multiple plugins are installed, possibly adding a `config` subcommand to the `cmd_plugin` command group in palm-cli.

Let's leave this as `dbt-config` for now. We can work on the palm-cli updates to support this and update later.

**Comment by jakeberesford-palmetto** [comment]
Adding no merge, this PR will be superseded by another one I'm working on. Leaving this open for now as a point of reference.

**Pull request closed without merging**